lib/neo4j/active_node/query/query_proxy_eager_loading.rb

```diff
@@ -2,58 +2,153 @@ module Neo4j
+            self.name = name
+            self.association = name ? model.associations[name] : nil
+          end
 
-          query_from_association_spec.pluck(identity, "[#{with_associations_return_clause}]").map do |record, eager_data|
-            eager_data.each_with_index do |eager_records, index|
-              record.association_proxy(with_associations_spec[index]).cache_result(eager_records)
+          def clone
+            super.tap { |copy| copy.each { |key, value| copy[key] = value.clone } }
+          end
+
+          def add_association(spec)
+            if spec.is_a?(Array)
+              spec.each { |s| add_association(s) }
+            elsif spec.is_a?(Hash)
+              spec.each { |k, v| (self[k] ||= AssociationTree.new(model, k)).add_association(v) }
+            else
+              self[spec] ||= AssociationTree.new(model, spec)
             end
+          end
+
+          def paths(*prefix)
+            values.flat_map { |v| [[*prefix, v]] + v.paths(*prefix, v) }
+          end
 
-            yield(record)
+          private
+
+          def target_class(model, key)
+            association = model.associations[key]
+            fail "Invalid association: #{[*path, key].join('.')}" unless association
+            model.associations[key].target_class
           end
         end
 
-        def with_associations_spec
-          @with_associations_spec ||= []
+        def pluck_vars(node, rel)
+          return super if with_associations_spec.size.zero?
+
+          perform_query
         end
 
-        def with_associations(*spec)
-          invalid_association_names = spec.reject do |association_name|
-            model.associations[association_name]
-          end
+        def perform_query
+          @_cache = IdentityMap.new
+          query_from_association_spec
+            .map do |record, eager_data|
+            cache_and_init(record, with_associations_spec)
+            eager_data.zip(with_associations_spec.paths.map(&:last)).each do |eager_records, element|
+              eager_records.first.zip(eager_records.last).each do |eager_record|
+                add_to_cache(*eager_record, element)
+              end
+            end
 
-          if !invalid_association_names.empty?
-            fail "Invalid associations: #{invalid_association_names.join(', ')}"
+            record
           end
+        end
 
+        def with_associations(*spec)
           new_link.tap do |new_query_proxy|
-            new_spec = new_query_proxy.with_associations_spec + spec
-            new_query_proxy.with_associations_spec.replace(new_spec)
+            new_query_proxy.with_associations_spec = with_associations_spec.clone
+            new_query_proxy.with_associations_spec.add_association(spec)
           end
         end
 
+        def with_associations_spec
+          @with_associations_spec ||= AssociationTree.new(model)
+        end
+
+        def with_associations_spec=(tree)
+          @with_associations_spec = tree
```

Comment:
Probably we can change all `with_associations_spec` references to something like `with_association_tree`